build/purger/purger.py

```diff
@@ -0,0 +1,235 @@
+        sys.exit(1)
+
+    api = GMatrixHttpApi(server)
+    try:
+        response = api.login("m.login.password", user=username, password=password)
+        admin_access_token = response["access_token"]
+    except (MatrixError, KeyError) as ex:
+        click.secho(f"Could not log in to server {server}: {ex}")
+        sys.exit(1)
+
+    try:
+        with psycopg2.connect(db_uri) as db, db.cursor() as cur:
+            purges: Dict[str, str] = dict()
+
+            def wait_and_purge_room(room_id: str = None, event_id: str = None) -> None:
+                """ Wait for available slots in parallel_purges and purge room
+
+                If room_id is None, just wait for current purges to complete and return
+                If event_id is None, purge all events in room
+                """
+                while len(purges) >= (parallel_purges if room_id else 1):
+                    # wait and clear completed purges
+                    time.sleep(1)
+                    for _room_id, purge_id in list(purges.items()):
+                        response = session.get(
+                            urljoin(
+                                server,
+                                "/_matrix/client/r0/admin/purge_history_status/" + quote(purge_id),
+                            ),
+                            params={"access_token": admin_access_token},
+                        )
+                        assert response.status_code == 200, f"{response!r} => {response.text!r}"
+                        if response.json()["status"] != "active":
+                            click.secho(f"Finished purge: room {_room_id!r}, purge {purge_id!r}")
+                            purges.pop(_room_id)
+
+                if not room_id:
+                    return
+
+                body: Dict[str, Any] = {"delete_local_events": True}
+                if event_id:
+                    body["purge_up_to_event_id"] = event_id
+                else:
+                    body["purge_up_to_ts"] = int(time.time() * 1000)
+                response = session.post(
+                    urljoin(server, "/_matrix/client/r0/admin/purge_history/" + quote(room_id)),
+                    params={"access_token": admin_access_token},
+                    json=body,
+                )
+                if response.status_code == 200:
+                    purge_id = response.json()["purge_id"]
+                    purges[room_id] = purge_id
+                    return
+
+            if not keep_newer and not keep_min_msgs:
+                click.confirm(
+                    "No --keep-newer nor --keep-min-msgs option provided. Purge all history?",
+                    abort=True,
+                )
+
+            ts_ms = None
+            if keep_newer:
+                ts = datetime.datetime.now() - datetime.timedelta(keep_newer)
+                ts_ms = int(ts.timestamp() * 1000)
+
+            cur.execute("SELECT room_id FROM rooms ;")
```

Comment:
Is there no API for removing rooms?

Reply:
This is getting a list of all rooms. I don't think there is an API endpoint that returns all rooms including non-public one.